programs/server/Server.cpp

```diff
@@ -1118,6 +1118,18 @@ int Server::main(const std::vector<std::string> & /*args*/)
             if (config->has("max_partition_size_to_drop"))
                 global_context->setMaxPartitionSizeToDrop(config->getUInt64("max_partition_size_to_drop"));
 
+            if (config->has("total_max_threads"))
+            {
+                auto total_max_threads = config->getInt("total_max_threads", 0);
+                if (total_max_threads == -1)
+                {
+                    // Based on tests total_max_threads has an optimal value when it's about two times of logical CPU cores
+                    constexpr size_t thread_factor = 2;
```

Comment:
This sounds small, since clickhouse uses threads for everything, for example each query has at least two threads, but they do not eats lots of CPU, so I would say that `thread_factor=10` (or maybe even 20) should be used at least by default.
What do you think?

Reply:
@azat , it is usually believed that the optimal number of active threads in a system is number_of_CPU_cores, or number_of_CPU_cores*N, where N is something like 2 or 4. Our experiments proved it.
N depends on the kind of CPU (particularly if hyperthreading or similar technology is enabled) and depends on delays caused by disks subsystem and by interactions with external entities, so in some exotic cases 10 might be better than 2, but 2 or 4 seems much better default values.

Reply:
Yes, if our tests were disk bound, result would be different.

Reply:
Not only disk bound, thread may just poll something once in a ms or so.

Reply:
Yes, this is true.
The worst scenario: number of disk or network bound queries that 'eat' total_max_threads, then a CPU bound query. As a result - low CPU utilization.
I think that it is not a problem of a bad thread_factor, it is just a limitation of this approach. BTW, initially, we thought about a more sophisticated way to calculate max_threads based on load average or something, but found out that total_max_threads works fine for our scenarios.
Summing up, I suggest

1. use 4, not 2 as _thread_factor_
2. make the comment in the source code more detailed, mentioning that default is good for CPU bound scenarios and maybe suboptimal for other ones
3. add an entry in the example config with long and comprehensive explanation how to estimate the proper value and when the default value is not perfect.

@azat ?

Reply:
>make the comment in the source code more detailed, mentioning that default is good for CPU bound scenarios and maybe suboptimal for other ones

Agree, better to add some comment.

>add an entry in the example config with long and comprehensive explanation how to estimate the proper value and when the default value is not perfect.

Yes, great.

>use 4, not 2 as thread_factor

To me, 4 looks too small.

Sane defaults is very important for massive usage, since very few users looks at config.xml/documentation to tune some settings, most of them uses default, and since not all threads in ClickHouse is CPU bound, I voting for 10 at least.

But if everybody else think that 4 is good, then I'm fine either, since I will either change the default or not using this limit at all.

Reply:
> To me, 4 looks too small.

> Sane defaults is very important for massive usage, since very few users looks at config.xml/documentation to tune some settings, most of them uses default, and since not all threads in ClickHouse is CPU bound, I voting for 10 at least.

At first glance it is strange to conduct massive tests (some are described in article.pdf attached to the PR), get clear results, and ignore these results afterward.

Reply:
Ok, thinking about this a little bit more and I'm convinced me that 4 is OK as default (or even 3, since in your pdf 2-3 is optimal AFAICS), since
- this is actually not total threads limit (and hence total_max_threads does not looks a good naming), but just a threshold after what only one thread will be used for query execution.
- plus this setting does not takes into account some threads (like TCPHandler thread)

Reply:
Agree with all your statements. 

I admit that total_max_threads is misleading. To me, the risk of misunderstanding should be mitigated by a comment, but maybe I am wrong and we really should think about a better name.

Reply:
@azat  I pushed a new commit where I set thread_factor = 3;